File: src/optic/lua/flat/ops/Operator.java
package optic.lua.flat.ops;

import optic.lua.flat.*;

public class Operator implements Step {
	private final StepType type;
	private final Register a, b, target;
	private final String symbol;

	public Operator(StepType type, Register a, Register b, Register target, String symbol) {
		this.type = type;
		this.a = a;
		this.b = b;
		this.target = target;
		this.symbol = symbol;
	}

	public Operator(StepType type, Register b, Register target, String symbol) {
		this.type = type;
		this.a = null;
		this.b = b;
		this.target = target;
		this.symbol = symbol;
	}

	@Override
	public StepType getType() {
		return type;
	}

	@Override
	public String toString() {
		return String.format("%s %s = %s %s %s",
				typeName(),
				target,
				(a != null ? a : ""), symbol, b);
	}
}

File: src/optic/lua/asm/instructions/Operator.kt
package optic.lua.asm.instructions

import optic.lua.asm.Register
import optic.lua.asm.Step
import java.util.*

class Operator : Step {
    val a: Register?
    val b: Register
    val target: Register
    val symbol: String

    constructor(a: Register, b: Register, target: Register, symbol: String) {
        this.a = a
        this.b = b
        this.target = target
        this.symbol = symbol
    }

    constructor(b: Register, target: Register, symbol: String) {
        this.a = null
        this.b = b
        this.target = target
        this.symbol = symbol
    }

    override fun toString(): String {
        return "op $target = ${a ?: ""} $symbol $b"
    }

    override fun modified(): Register? = target
}
